common/io.cpp

```diff
@@ -73,6 +80,79 @@ ExpectedSize StreamWriter::Write(
 	return end - start;
 }
 
+class ReaderStreamBuffer : public streambuf {
+public:
+	ReaderStreamBuffer(Reader &reader) :
+		reader_ {reader},
+		buf_(buf_size_) {};
+	streambuf::int_type underflow() override;
+
+private:
+	static const Vsize buf_size_ = 1024;
+	Reader &reader_;
+	vector<uint8_t> buf_;
+};
+
+streambuf::int_type ReaderStreamBuffer::underflow() {
+	// eback -- pointer to the first char (byte)
+	// gptr  -- pointer to the current char (byte)
+	// egptr -- pointer past the last char (byte)
+
+	// This function is only called if gptr() == nullptr or gptr() >= egptr(),
+	// i.e. if there's nothing more to read.
+	if (this->gptr() >= this->egptr()) {
+		auto ex_n_read = reader_.Read(buf_.begin(), buf_.end());
+		streamsize n_read;
+		if (ex_n_read) {
+			n_read = ex_n_read.value();
+		} else {
+			// There is no way to return an error from underflow(), generally
+			// the streams only care about how much data was read. No data or
+			// less data then requested by the caller of istream.read() means
+			// eofbit and failbit are set. If the user code wants to get the
+			// error or check if there was an error, it needs to check errno.
+			//
+			// So as long as we don't clear errno after a failure in the
+			// reader_.Read() above, error handling works as usual and returning
+			// eof below is all that needs to happen here.
+			n_read = 0;
+		}
+
+		streambuf::char_type *first = reinterpret_cast<streambuf::char_type *>(buf_.data());
+
+		// set eback, gptr, egptr
+		this->setg(first, first, first + n_read);
+	}
+
+	return this->gptr() == this->egptr() ? std::char_traits<char>::eof()
+										 : std::char_traits<char>::to_int_type(*this->gptr());
```

Comment:
`clang-format` agrees with this..? :neutral_face:

Reply:
You think it was me formatting the code like this?!